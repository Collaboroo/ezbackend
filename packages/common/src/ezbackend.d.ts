/// <reference types="node" />
import { EzApp, EzBackendServer } from "./ezapp";
import fastify, { FastifyInstance } from "fastify";
import { Connection, createConnection, EntitySchema, ObjectLiteral, Repository } from "typeorm";
import { InjectOptions } from "light-my-request";
export interface EzBackendInstance {
    entities: Array<EntitySchema>;
    server: EzBackendServer;
    _server: FastifyInstance;
    repo: Repository<ObjectLiteral>;
    orm: Connection;
}
export interface EzBackendOpts {
    address: string;
    port: number;
    orm: Parameters<typeof createConnection>[0];
    server: Parameters<typeof fastify>[0];
}
/**
 * Child of EzApp. This is where you set up your backend setup tasks.
 */
export declare class EzBackend extends EzApp {
    constructor();
    getInternalInstance(): EzBackendInstance;
    getInternalServer(): FastifyInstance<import("http").Server, import("http").IncomingMessage, import("http").ServerResponse, import("fastify").FastifyLoggerInstance>;
    inject(injectOpts: string | InjectOptions): Promise<import("light-my-request").Response>;
    verifyStarted(funcName?: string): void;
    printRoutes(): string;
    printPlugins(): string;
    prettyPrint(): string;
    start(opts?: any): Promise<void>;
}
