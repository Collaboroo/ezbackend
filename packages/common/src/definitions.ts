import {EzBackend} from '@ezbackend/core'
import {Connection, ConnectionOptions} from 'typeorm'
import {FastifyInstance, FastifyLoggerOptions} from "fastify"
import { EzModelMeta } from '.'

export interface IOptions {
    server?: {
        port?: number
        logger?:FastifyLoggerOptions
    }
    orm?: ConnectionOptions
    
}

declare module '@ezbackend/core' {
    interface EzBackend {
        orm: Connection,
        server: FastifyInstance
        models: Array<EzModelMeta>
    }
}

