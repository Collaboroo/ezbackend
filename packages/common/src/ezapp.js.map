{"version":3,"file":"ezapp.js","sourceRoot":"","sources":["ezapp.ts"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,wCAAkD;AAElD,kEAA+B;AAS/B,SAAS,0BAA0B,CAC/B,MAAa,EACb,QAAyC;IAGzC,OAAO;QAAC,cAAe;aAAf,UAAe,EAAf,qBAAe,EAAf,IAAe;YAAf,yBAAe;;QACnB,MAAM,CAAC,SAAS,CAAC,IAAI;QACjB,+DAA+D;QAC/D,UAAC,MAAuB,IAAK,OAAC,MAAM,CAAC,QAAQ,CAAC,OAAhB,MAAM,EAAsB,IAAI,GAAjC,CAAkC,CAClE,CAAA;IACL,CAAC,CAAA;AACL,CAAC;AAED,2CAA2C;AAC3C,wCAAwC;AACxC,SAAS,YAAY,CAAC,MAAa;IAC/B,OAAO;QACH,kDAAkD;QAClD,QAAQ;QACR,MAAM,EAAE,0BAA0B,CAAoD,MAAM,EAAE,QAAQ,CAAC;QACvG,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,IAAI,EAAE,0BAA0B,CAAkD,MAAM,EAAE,MAAM,CAAC;QACjG,KAAK,EAAE,0BAA0B,CAAmD,MAAM,EAAE,OAAO,CAAC;QACpG,IAAI,EAAE,0BAA0B,CAAkD,MAAM,EAAE,MAAM,CAAC;QACjG,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,OAAO,EAAE,0BAA0B,CAAqD,MAAM,EAAE,SAAS,CAAC;QAC1G,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,KAAK,EAAE,0BAA0B,CAAmD,MAAM,EAAE,OAAO,CAAC;QACpG,mJAAmJ;QACnJ,OAAO,EAAE,0BAA0B,CAAmD,MAAM,EAAE,SAAS,CAAC;QACxG,SAAS,EAAE,0BAA0B,CAAmD,MAAM,EAAE,WAAW,CAAC;QAC5G,qBAAqB,EAAE,0BAA0B,CAAmE,MAAM,EAAE,uBAAuB,CAAC;QACpJ,oBAAoB,EAAE,0BAA0B,CAAkE,MAAM,EAAE,sBAAsB,CAAC;QACjJ,QAAQ,EAAE,0BAA0B,CAAsD,MAAM,EAAE,UAAU,CAAC;QAC7G,aAAa,EAAE,0BAA0B,CAA2D,MAAM,EAAE,eAAe,CAAC;QAC5H,eAAe,EAAE,0BAA0B,CAA6D,MAAM,EAAE,iBAAiB,CAAC;QAClI,MAAM,EAAE,0BAA0B,CAAgD,MAAM,EAAE,QAAQ,CAAC;QACnG,6EAA6E;QAC7E,QAAQ,EAAE,0BAA0B,CAAM,MAAM,EAAE,UAAU,CAAoB;QAChF,kBAAkB,EAAE,0BAA0B,CAAgE,MAAM,EAAE,oBAAoB,CAAC;QAC3I,eAAe,EAAE,0BAA0B,CAA6D,MAAM,EAAE,iBAAiB,CAAC;KAErI,CAAA;AACL,CAAC;AAID;;;GAGG;AACH;IAA2B,yBAAG;IAM1B;;OAEG;IACH;QAAA,YACI,iBAAO,SASV;QAjBS,gBAAU,GAAoB,EAAE,CAAA;QAmB1C,6BAA6B;QAC7B,kFAAkF;QAClF,YAAM,GAAG,0BAA0B,CAAoD,KAAI,EAAE,QAAQ,CAAC,CAAA;QACtG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,UAAI,GAAG,0BAA0B,CAAkD,KAAI,EAAE,MAAM,CAAC,CAAA;QAChG,WAAK,GAAG,0BAA0B,CAAmD,KAAI,EAAE,OAAO,CAAC,CAAA;QACnG,UAAI,GAAG,0BAA0B,CAAkD,KAAI,EAAE,MAAM,CAAC,CAAA;QAChG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,aAAO,GAAG,0BAA0B,CAAqD,KAAI,EAAE,SAAS,CAAC,CAAA;QACzG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,WAAK,GAAG,0BAA0B,CAAmD,KAAI,EAAE,OAAO,CAAC,CAAA;QACnG,aAAO,GAAG,0BAA0B,CAAM,KAAI,EAAE,SAAS,CAA+B,CAAA;QACxF,eAAS,GAAG,0BAA0B,CAAmD,KAAI,EAAE,WAAW,CAAC,CAAA;QAC3G,2BAAqB,GAAG,0BAA0B,CAAmE,KAAI,EAAE,uBAAuB,CAAC,CAAA;QACnJ,0BAAoB,GAAG,0BAA0B,CAAkE,KAAI,EAAE,sBAAsB,CAAC,CAAA;QAChJ,cAAQ,GAAG,0BAA0B,CAAsD,KAAI,EAAE,UAAU,CAAC,CAAA;QAC5G,mBAAa,GAAG,0BAA0B,CAA2D,KAAI,EAAE,eAAe,CAAC,CAAA;QAC3H,qBAAe,GAAG,0BAA0B,CAA6D,KAAI,EAAE,iBAAiB,CAAC,CAAA;QAEjI,cAAQ,GAAG,0BAA0B,CAAsD,KAAI,EAAE,UAAU,CAAC,CAAA,CAAC,yCAAyC;QACtJ,wBAAkB,GAAG,0BAA0B,CAAgE,KAAI,EAAE,oBAAoB,CAAC,CAAA;QAC1I,qBAAe,GAAG,0BAA0B,CAA6D,KAAI,EAAE,iBAAiB,CAAC,CAAA;QACjI,mEAAmE;QACnE,sDAAsD;QAEtD,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACvH,aAAO,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,OAAO,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACjH,iBAAW,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,WAAW,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACzH,mBAAa,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,aAAa,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC7H,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACvH,oBAAc,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,cAAc,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC/H,eAAS,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,SAAS,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACrH,YAAM,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,MAAM,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC/G,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QA3CnH,KAAI,CAAC,UAAU,CAAC,oBAAoB,EAAE,UAAO,QAAQ,EAAE,IAAI;;gBACvD,QAAQ,CAAC,MAAM,GAAG,YAAY,CAAC,IAAI,CAAC,CAAA;;;aACvC,CAAC,CAAA;QACF,KAAI,CAAC,cAAc,CAAC,oBAAoB,EAAE,UAAO,QAAQ,EAAE,IAAI;;gBAC3D,uGAAuG;gBACvG,YAAY;gBACZ,OAAO,QAAQ,CAAC,MAAM,CAAA;;;aACzB,CAAC,CAAA;;IACN,CAAC;IAfD,sBAAI,4BAAS;aAAb,cAAkB,OAAO,IAAI,CAAC,UAAU,CAAA,CAAC,CAAC;;;OAAA;IAoD1C;;;;OAIG;IACH,sCAAsB,GAAtB,UAAuB,MAAuB,EAAE,MAAa;QAA7D,iBAiBC;QAfG,IAAM,SAAS,GAAG,UAAO,MAAuB;;gBAC5C,MAAM,CAAC,SAAS,CAAC,OAAO,CAAC,UAAA,IAAI;oBACzB,IAAI,CAAC,MAAM,CAAC,CAAA;gBAChB,CAAC,CAAC,CAAA;gBACF,MAAM,CAAC,IAAI,CAAC,OAAO,CAAC,UAAA,GAAG;oBACnB,IAAI,GAAG,YAAY,KAAK,EAAE;wBACtB,GAAG,CAAC,sBAAsB,CAAC,MAAM,EAAE,GAAG,CAAC,CAAA;qBAC1C;gBACL,CAAC,CAAC,CAAA;;;aACL,CAAA;QAED,+GAA+G;QAC/G,IAAM,eAAe,GAAG,CAAC,MAAM,CAAC,KAAK,KAAK,kBAAW,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC,IAAA,wBAAE,EAAC,SAAS,CAAC,CAAC,CAAC,CAAC,SAAS,CAAA;QAEzF,MAAM,CAAC,QAAQ,CAAC,eAAe,EAAE,MAAM,CAAC,IAAI,CAAC,CAAA;IACjD,CAAC;IACL,YAAC;AAAD,CAAC,AA/ED,CAA2B,UAAG,GA+E7B;AA/EY,sBAAK"}