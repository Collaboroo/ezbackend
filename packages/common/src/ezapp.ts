import { App } from "@ezbackend/core"

function generateFastifyFuncWrapper(parent,funcName:string) {

    return (...opts) => {
        parent._functions.push(
            (server) => server[funcName](...opts)
        )
    }
}

function createServer(parent:EzApp) {
    return {
        route: generateFastifyFuncWrapper(parent,'route'),
        register: generateFastifyFuncWrapper(parent,'register'),
        addSchema: generateFastifyFuncWrapper(parent,'addSchema'),
        decorateReply: generateFastifyFuncWrapper(parent,'decorateReply'),
        inject: generateFastifyFuncWrapper(parent,'inject'),
        addHook: generateFastifyFuncWrapper(parent,'addHook'),
    }
}

export class EzApp extends App {

    protected _functions: Array<Function> = []

    get functions() { return this._functions }

    constructor() {
        super()
        this.setHandler("Create Server Stub", async (instance, opts) => {
            instance.server = createServer(this)
        })
        this.setPostHandler("Remove Server Stub", async (instance, opts) => {
            delete instance.server
        })

    }

    registerFastifyPlugins(server,parent) {

        server.register(async (server, opts) => {
            parent.functions.forEach(func => {
                func(server)
            })
            parent.apps.forEach(app => {
                if (app instanceof EzApp) {
                    app.registerFastifyPlugins(server,app)
                }
            })
        }, parent.opts)
    }
}