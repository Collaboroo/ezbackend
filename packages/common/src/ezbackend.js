"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EzBackend = void 0;
var ezapp_1 = require("./ezapp");
var fastify_1 = __importDefault(require("fastify"));
var fastify_plugin_1 = __importDefault(require("fastify-plugin"));
var typeorm_1 = require("typeorm");
var core_1 = require("@ezbackend/core");
var utils_1 = require("@ezbackend/utils");
var lodash_1 = __importDefault(require("lodash"));
var path_1 = __importDefault(require("path"));
var dotenv_1 = __importDefault(require("dotenv"));
//TODO: Check if emojis will break instance names
//URGENT TODO: Strict types for instance, opts
function addErrorSchema(instance, opts) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            instance.server.addSchema({
                "$id": "ErrorResponse",
                type: 'object',
                properties: {
                    statusCode: { type: 'number' },
                    error: { type: 'string' },
                    message: { type: 'string' }
                }
            });
            return [2 /*return*/];
        });
    });
}
//URGENT TODO: Make running this optional in the default config
dotenv_1.default.config();
var defaultConfig = {
    port: process.env.PORT || 8000,
    address: process.env.ADDRESS || "127.0.0.1",
    server: {
        logger: {
            prettyPrint: {
                translateTime: "SYS:HH:MM:ss",
                ignore: "pid,hostname,reqId,responseTime,req,res",
                messageFormat: "[{req.method} {req.url}] {msg}",
            },
        },
    },
    orm: {
        type: "better-sqlite3",
        database: ":memory:",
        synchronize: true
    },
    auth: {
        secretKey: (_a = process.env.SECRET_KEY) !== null && _a !== void 0 ? _a : undefined,
        secretKeyPath: path_1.default.join(process.cwd(), 'secret-key'),
        successRedirectURL: "http://localhost:8000/db-ui",
        failureRedirectURL: "http://localhost:8000/db-ui",
        google: {
            googleClientId: process.env.GOOGLE_CLIENT_ID,
            googleClientSecret: process.env.GOOGLE_CLIENT_SECRET,
            scope: ['profile'],
        }
    },
    // cors: {
    //     origin: (origin: string, cb: Function) => {
    //         if (/localhost/.test(origin)) {
    //             //  Request from localhost will pass
    //             cb(null, true)
    //             return
    //         }
    //         // Generate an error on other origins, disabling access
    //         cb(new Error("Not allowed"))
    //     }
    // }
};
// Derived from https://github.com/jeromemacias/fastify-boom/blob/master/index.js
// Kudos to him
var ezbErrorPage = function (fastify, options, next) {
    //TODO: Strict types for error
    fastify.setErrorHandler(function errorHandler(error, request, reply) {
        request.log.error(error);
        if (error && error.query) {
            //Assumption: It is a typeorm error if it falls here
            request.log.error("query: " + error.query);
            request.log.error("parameters: " + error.parameters);
            request.log.error("driverError: " + error.driverError);
        }
        if (error && error.isBoom) {
            reply
                .code(error.output.statusCode)
                .type('application/json')
                .headers(error.output.headers)
                .send(error.output.payload);
            return;
        }
        reply.send(error || new Error("Got non-error: " + error));
    });
    next();
};
/**
 * Child of EzApp. This is where you set up your backend setup tasks.
 */
var EzBackend = /** @class */ (function (_super) {
    __extends(EzBackend, _super);
    function EzBackend() {
        var _this = _super.call(this) || this;
        _this.setInit('Create Entities Container', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                instance.entities = [];
                return [2 /*return*/];
            });
        }); });
        _this.setPostInit('Create Database Connection', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = instance;
                        return [4 /*yield*/, (0, typeorm_1.createConnection)(__assign(__assign({}, opts.orm), { entities: instance.entities }))];
                    case 1:
                        _a.orm = _b.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        _this.setHandler('Add Fastify Boom', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                instance.server.register((0, fastify_plugin_1.default)(ezbErrorPage));
                return [2 /*return*/];
            });
        }); });
        _this.setHandler('Add Error Schema', addErrorSchema);
        _this.setPostHandler('Create Fastify Server', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                instance._server = (0, fastify_1.default)(opts.server);
                return [2 /*return*/];
            });
        }); });
        _this.setPostHandler('Register Fastify Plugins', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.registerFastifyPlugins(instance._server, this);
                return [2 /*return*/];
            });
        }); });
        _this.setRun('Run Fastify Server', function (instance, opts) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, instance._server.listen(opts.port, opts.address)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        _this.scope = core_1.PluginScope.PARENT;
        return _this;
    }
    EzBackend.prototype.getInternalInstance = function () {
        //TODO: Figure if there is a better way of getting this data
        //@ts-ignore
        var lastPlugin = this.instance._lastUsed;
        if (lastPlugin === null) {
            throw new Error("Server is still undefined, have you called app.start() yet?");
        }
        return lastPlugin.server;
    };
    EzBackend.prototype.getInternalServer = function () {
        return this.getInternalInstance()._server;
    };
    EzBackend.prototype.inject = function (injectOpts) {
        return __awaiter(this, void 0, void 0, function () {
            var server;
            return __generator(this, function (_a) {
                server = this.getInternalServer();
                return [2 /*return*/, server.inject(injectOpts)];
            });
        });
    };
    EzBackend.prototype.verifyStarted = function (funcName) {
        if (!this.instance.started) {
            var additionalMsg = funcName
                ? "before running " + funcName
                : '';
            throw new utils_1.EzError("Instance not yet started", "The EzBackend instance must be started " + additionalMsg, "\nawait app.start()\n\nYou must wait for the above function to finish before you can run " + funcName + "\n");
        }
    };
    EzBackend.prototype.printRoutes = function () {
        this.verifyStarted("printRoutes");
        return this.getInternalServer().printRoutes();
    };
    EzBackend.prototype.printPlugins = function () {
        this.verifyStarted("printPlugins");
        return this.getInternalServer().printPlugins();
    };
    EzBackend.prototype.prettyPrint = function () {
        this.verifyStarted("prettyPrint");
        return this.instance.prettyPrint();
    };
    //URGENT TODO: Remove temporary any fix
    EzBackend.prototype.start = function (opts) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        opts = lodash_1.default.merge(defaultConfig, opts);
                        return [4 /*yield*/, _super.prototype.start.call(this, opts)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return EzBackend;
}(ezapp_1.EzApp));
exports.EzBackend = EzBackend;
//# sourceMappingURL=ezbackend.js.map