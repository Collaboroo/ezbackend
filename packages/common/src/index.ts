import { IOptions } from "./definitions";
import { IEzbConfig } from "@ezbackend/core";
import { fastify } from "fastify";
import path from "path";
import { EzBackend } from "./definitions";
import { mixedInstance } from "avvio";
import { createConnection } from "typeorm";
import { RouteOptions } from "fastify";
import { APIGenerator } from "./models";
import {convert} from './models/typeorm-json-schema'

const ezb = EzBackend.app();

//Configure defaults
ezb.plugins.init = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig & IOptions,
  cb
) => {
  ezb.orm = await createConnection(opts.orm);
  ezb.server = fastify(opts.server);
  cb();
};

ezb.plugins.handler = (ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  const customEzbPath =
    opts.entryPoint ?? path.join(process.cwd(), ".ezb/index.ts");
  require(customEzbPath);
  ezb.models.forEach((model) => {
    const repo = ezb.orm.getRepository(model);

    //LEFT OFF
    // const metaData = ezb.orm.getMetadata(model)
    // console.log(convert(metaData))

    const generator = new APIGenerator(repo, { prefix: repo.metadata.name });
    generator.generateRoutes();
  });
  cb();
};

ezb.plugins.run = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig,
  cb
) => {
  await ezb.server.listen(opts.port, function (err, address) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
  cb();
};

export { EzBackend } from "./definitions";
export { EzModel, response, APIGenerator } from "./models";
