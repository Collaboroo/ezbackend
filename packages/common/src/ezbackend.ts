import { EzApp, EzBackendServer } from "./ezapp";
import fastify, { FastifyInstance } from "fastify";
import fastifyBoom from 'fastify-boom'
import { Connection, createConnection, EntitySchema, ObjectLiteral, Repository } from "typeorm";
import { PluginScope } from "@ezbackend/core";
import _ from 'lodash'
import path from 'path'
import dotenv from 'dotenv'
import { InjectOptions } from "light-my-request";

export interface EzBackendInstance {
    entities: Array<EntitySchema>
    server: EzBackendServer
    _server: FastifyInstance
    repo: Repository<ObjectLiteral>
    orm: Connection
}

export interface EzBackendOpts {
    port: number
    orm: Parameters<typeof createConnection>[0]
    server: Parameters<typeof fastify>[0]
}

//TODO: Check if emojis will break instance names
//URGENT TODO: Strict types for instance, opts
async function addErrorSchema(instance: EzBackendInstance, opts: EzBackendOpts) {
    instance.server.addSchema({
        "$id": "ErrorResponse",
        type: 'object',
        properties: {
            statusCode: { type: 'number' },
            error: { type: 'string' },
            message: { type: 'string' }
        }
    })
}

//URGENT TODO: Make running this optional in the default config
dotenv.config()

const defaultConfig = {
    port: 8000,
    server: {
        logger: {
            prettyPrint: {
                translateTime: "SYS:HH:MM:ss",
                ignore: "pid,hostname,reqId,responseTime,req,res",
                messageFormat: "[{req.method} {req.url}] {msg}",
            },
        },
    },
    orm: {
        type: "better-sqlite3",
        database: ":memory:",
        synchronize: true
    },
    auth: {
        secretKeyPath: path.join(process.cwd(), 'secret-key'),
        successRedirectURL: "http://localhost:8000/db-ui",
        failureRedirectURL: "http://localhost:8000/db-ui",
        google: {
            googleClientId: process.env.GOOGLE_CLIENT_ID,
            googleClientSecret: process.env.GOOGLE_CLIENT_SECRET,
            scope: ['profile'],
        }
    },
    // cors: {
    //     origin: (origin: string, cb: Function) => {
    //         if (/localhost/.test(origin)) {
    //             //  Request from localhost will pass
    //             cb(null, true)
    //             return
    //         }
    //         // Generate an error on other origins, disabling access
    //         cb(new Error("Not allowed"))
    //     }
    // }
}

/**
 * Child of EzApp. This is where you set up your backend setup tasks.
 */
export class EzBackend extends EzApp {

    constructor() {
        super()

        this.setInit('Create Entities Container', async (instance, opts) => {
            instance.entities = []
        })
        this.setPostInit('Create Database Connection', async (instance, opts) => {
            instance.orm = await createConnection(
                {
                    ...opts.orm,
                    entities: instance.entities
                }
            )
        })

        this.setHandler('Add Fastify Boom', async (instance, opts) => {
            instance.server.register(fastifyBoom)
        })
        this.setHandler('Add Error Schema', addErrorSchema)

        this.setPostHandler('Create Fastify Server', async (instance, opts) => {
            instance._server = fastify(opts.server)
        })

        this.setPostHandler('Register Fastify Plugins', async (instance, opts) => {
            this.registerFastifyPlugins(instance._server, this)
        })

        this.setRun('Run Fastify Server', async (instance, opts) => {
            await instance._server.listen(opts.port)
        })

        this.scope = PluginScope.PARENT

    }

    getInternalInstance() {
        //TODO: Figure if there is a better way of getting this data
        //@ts-ignore
        const lastPlugin = this.instance._lastUsed
        if (lastPlugin === null) {
            throw new Error("Server is still undefined, have you called app.start() yet?")
        }
        return lastPlugin.server as EzBackendInstance
    }

    getInternalServer() {
        return this.getInternalInstance()._server
    }

    async inject(injectOpts: string | InjectOptions) {
        const server = this.getInternalServer()
        return server.inject(injectOpts)
    }

    printRoutes() {
        return this.getInternalServer().printRoutes()
    }

    printPlugins() {
        return this.getInternalServer().printPlugins()
    }

    //URGENT TODO: Remove temporary any fix
    async start(opts?: any) {
        opts = _.merge(defaultConfig, opts)
        await super.start(opts)
    }

}