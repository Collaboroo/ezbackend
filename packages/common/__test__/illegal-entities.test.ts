import { EzBackend, EzBackendOpts, EzModel, RecursivePartial, Type } from "../src"
import { afterAll, beforeAll, describe, expect, test } from "@jest/globals";

import { EzError } from "@ezbackend/utils";

const defaultConfig: RecursivePartial<EzBackendOpts> = {
    ezbackend: {
        fastify: {
            logger: false
        },
        typeorm: {
            database: ':memory:'
        }
    }
}
describe("Illegal Entity Creation", () => {

    let app: EzBackend

    beforeAll(async () => {
        app = new EzBackend()
        app.removeHook("_run", "Run Fastify Server")

    })

    test("Creating a relation without the full syntax should be illegal", async () => {
        let errored = false

        try {
            app.addApp(new EzModel("IllegalModel", {
                relation: Type.ONE_TO_ONE
            }))
            await app.start(defaultConfig)

        } catch (e: unknown) {
            const err = e as EzError
            expect(err.summary).toMatchSnapshot()
            expect(err.description).toMatchSnapshot()
            expect(err.code).toMatchSnapshot()
            errored = true
        }

        expect(errored).toBe(true)
    })

    test("Creating Model with more than one primary column is illegal", async () => {
        let errored = false

        try {
            app.addApp(new EzModel("IllegalModel", {
                mySecondPrimaryColumn: {
                    type: Type.VARCHAR,
                    primary: true
                }
            }))
            await app.start(defaultConfig)

        } catch (e) {
            expect(e).toMatchSnapshot()
            errored = true
        }

        expect(errored).toBe(true)
    })

});
