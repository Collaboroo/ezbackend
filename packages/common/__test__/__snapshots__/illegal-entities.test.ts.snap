// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Illegal Entity Creation Creating Model with more than one primary column is illegal 1`] = `[Error: App has been started already!]`;

exports[`Illegal Entity Creation Creating a relation without the full syntax should be illegal 1`] = `"You currently need to use the full declaration for specifying a relation"`;

exports[`Illegal Entity Creation Creating a relation without the full syntax should be illegal 2`] = `"Relations require additional metadata to generate the Database Tables"`;

exports[`Illegal Entity Creation Creating a relation without the full syntax should be illegal 3`] = `
"Replace

myRelation: Type.ONE_TO_ONE

With

myRelation: {
    type: Type.ONE_TO_ONE,
    joinColumn: true,
    target:'detail'
},"
`;
